libraries/discriminator/syn/src/lib.rs

```diff
@@ -7,74 +7,43 @@ mod error;
     solana_program::hash,
-    syn::{parse::Parse, Generics, Ident, Item, ItemEnum, ItemStruct, LitByteStr, WhereClause},
+    syn::{ext::IdentExt, parse::Parse, Attribute, Ident, LitByteStr, Token, Visibility},
 };
 
 /// "Builder" struct to implement the `SplDiscriminate` trait
 /// on an enum or struct
 pub struct SplDiscriminateBuilder {
     /// The struct/enum identifier
     pub ident: Ident,
-    /// The item's generic arguments (if any)
-    pub generics: Generics,
-    /// The item's where clause for generics (if any)
-    pub where_clause: Option<WhereClause>,
     /// The TLV hash_input
     pub hash_input: String,
 }
 
-impl TryFrom<ItemEnum> for SplDiscriminateBuilder {
-    type Error = SplDiscriminateError;
-
-    fn try_from(item_enum: ItemEnum) -> Result<Self, Self::Error> {
-        let ident = item_enum.ident;
-        let where_clause = item_enum.generics.where_clause.clone();
-        let generics = item_enum.generics;
-        let hash_input = parse_hash_input(&item_enum.attrs)?;
-        Ok(Self {
-            ident,
-            generics,
-            where_clause,
-            hash_input,
-        })
-    }
-}
-
-impl TryFrom<ItemStruct> for SplDiscriminateBuilder {
-    type Error = SplDiscriminateError;
-
-    fn try_from(item_struct: ItemStruct) -> Result<Self, Self::Error> {
-        let ident = item_struct.ident;
-        let where_clause = item_struct.generics.where_clause.clone();
-        let generics = item_struct.generics;
-        let hash_input = parse_hash_input(&item_struct.attrs)?;
-        Ok(Self {
-            ident,
-            generics,
-            where_clause,
-            hash_input,
-        })
-    }
-}
-
 impl Parse for SplDiscriminateBuilder {
     fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
-        let item = Item::parse(input)?;
-        match item {
-            Item::Enum(item_enum) => item_enum.try_into(),
-            Item::Struct(item_struct) => item_struct.try_into(),
-            _ => {
-                return Err(syn::Error::new(
-                    Span::call_site(),
-                    "Only enums and structs are supported",
-                ))
-            }
+        let attrs = input.call(Attribute::parse_outer)?;
+        let hash_input = parse_hash_input(&attrs)
+            .map_err(|e| syn::Error::new(input.span(), format!("Failed to parse item: {}", e)))?;
+        if input.peek(Token![pub]) {
+            input.parse::<Visibility>()?;
         }
-        .map_err(|e| syn::Error::new(input.span(), format!("Failed to parse item: {}", e)))
+        if input.peek(Token![struct]) ||input.peek(Token![enum]) {
+            input.call(Ident::parse_any)?;
+        }
+        let ident = input.parse::<Ident>()?;
+        // just consume the rest
+        input.step(|cursor| {
+            let mut rest = *cursor;
+            while let Some((_, next)) = rest.token_tree() {
+                rest = next;
+            }
+            Ok(((), rest))
+        })?;
```

Comment:
Once we have the `Ident` can't we just return?

Reply:
Nope! `parse_macro_input` will barf if the whole thing isn't consumed, saying `{` wasn't expected.